packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1228 @@
+    this.backButtonDispatcher,
+  }) : assert(routeInformationProvider == null || routeInformationParser != null),
+       assert(routerDelegate != null),
+       super(key: key);
+
+  /// The route information provider for the router.
+  ///
+  /// The value at the time of first build will be used as the initial route.
+  /// The [Router] listens to this provider and rebuilds with new names when
+  /// it notifies.
+  ///
+  /// This can be null if this router does not rely on the route information
+  /// to build its content. In such case, the [routeInformationParser] can also be
+  /// null.
+  final RouteInformationProvider routeInformationProvider;
+
+  /// The route information parser for the router.
+  ///
+  /// When the [Router] gets a new route information from the [routeInformationProvider],
+  /// the [Router] uses this delegate to parse the route information and produce a
+  /// configuration. The configuration will be used by [routerDelegate] and
+  /// eventually rebuilds the [Router] widget.
+  ///
+  /// Since this delegate is the primary consumer of the [routeInformationProvider],
+  /// it must not be null if [routeInformationProvider] is not null.
+  final RouteInformationParser<T> routeInformationParser;
+
+  /// The router delegate for the router.
+  ///
+  /// This delegate consumes the configuration from [routeInformationParser] and
+  /// builds a navigating widget for the [Router].
+  ///
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// The back button dispatcher for the router.
+  ///
+  /// The two common alternatives are the [RootBackButtonDispatcher] for root
+  /// router, or the [ChildBackButtonDispatcher] for other routers.
+  final BackButtonDispatcher backButtonDispatcher;
+
+  /// Retrieves the immediate [Router] ancestor from the given context.
+  ///
+  /// Use this method when you need to access the delegates in the [Router].
+  /// For example, you need to access the [backButtonDispatcher] of the parent
+  /// router to create a [ChildBackButtonDispatcher] for a nested router.
+  /// Another use case may be updating the value in [routeInformationProvider]
+  /// to navigate to a new route.
+  static Router<dynamic> of(BuildContext context) {
+    final _RouterScope scope = context.dependOnInheritedWidgetOfExactType<_RouterScope>();
+    assert(scope != null);
+    return scope.routerState.widget;
+  }
+
+  /// Forces the [Router] to to run the [callback] and reports the route
+  /// information back to the engine.
+  ///
+  /// Normally, calls to this method are not necessary because the [Router] will
+  /// reports route information back to the engine automatically if it changes.
```

Comment:
The docs should tell me when its not reported automatically and I need to use this method.